### ISSUE TITLE
browse.databaseFilename/${workspaceFolderBasename} fails when subdirectory isn't created first

### ISSUE THREAD

**Issue by boocs** [microsoft/vscode-cpptools#5181]
# State

### State: Global User settings.json

```
    "C_Cpp.default.browse.databaseFilename": "k:\\.vscode\\cache\\browse\\${workspaceFolderBasename}\\global.vs.db"
```
1. Setting: C_Cpp.default.browse.databaseFilename
2. Value: "k:\\\\.vscode\\\\cache\\\\browse\\\\${workspaceFolderBasename}\\\\global.vs.db"

### State: Cache directory
* k:\\\\.vscode\\\\cache
* No subdirectories except for ipch (intellisense cache)
* k:\\\\.vscode\\\\cache\\\\browse isn’t created yet

### State: System
* Windows 10 (latest)
* VSCode (1.43.1)
* Build Tools for Visual Studio 2019 (latest)

### State: Extensions
* cpptools | ms- | 0.26.3

 # Results
    It works, but you have to start VSCode twice:

### Results: First Startup
1. VSCode creates k:\\\\.vscode\\\\cache\\\\browse
2. \\\\${workspaceFolderBasename}\\\\global.vs.db is not created
3. C++ Debug console: 
```
Failed to open the file k:\.vscode\cache\browse\MyApp\global.vs.db
```
4. Tag Parser doesn’t work

### Results: Second Startup
1. ${workspaceFolderBasename}\\\\global.vs.db gets created now that k:\\\\.vscode\\\\cache\\\\browse subdirectory has been created (on the first start of VSCode)
2. OS shows correct path has been created: k:.vscode\\\\cache\\\\browse\\\\MyApp\\\\global.vs.db
3. **Tag Parser works as expected**

# Irrelevant State (Probably)

### State: Global User settings.json
```
  "C_Cpp.intelliSenseCachePath": "k:\\.vscode\\cache",
  "C_Cpp.default.browse.path": [ ],
  "C_Cpp.default.browse.limitSymbolsToIncludedHeaders": true,
  "C_Cpp.default.compilerPath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.25.28610\\bin\\Hostx64\\x64\\cl.exe",
  "C_Cpp.loggingLevel": "Debug"
```

**Comment by Colengms** [comment]
@boocs Are you still seeing this issue?  I suspect this has since been addressed, as I'm not able to reproduce the issue in a recent version of the extension.

**Comment by boocs** [comment]
> @boocs Are you still seeing this issue? I suspect this has since been addressed, as I'm not able to reproduce the issue in a recent version of the extension.

Yep the bug still happens. The only difference is I don't get the log message 'Failed to open the file' anymore. 
Opening twice also still fixes it.

I also tried moving it to my C: drive and same thing happens.

I'm  using v1.0.1 (Sept 21)

**Comment by Colengms** [comment]
Hi @boocs.  Perhaps there is some additional information needed about your repro.  In my case, I have a drive Z:, with no ".vscode" directory, and when I set `C_Cpp.default.browse.databaseFilename` to `Z:\\.vscode\\cache\\browse\\${workspaceFolderBasename}\\global.vs.db`, I immediately see the entire tree of folders created, and see that the extension start tag parsing into the new database.  I'm not seeing partial creation of directories.  Since those directories are created one after another, from top to bottom, in a simple loop, that behavior would seem to suggest that we successfully created some, until one failed.  And, that after a while, we were able to create the one that had previously failed.  Is there something special about your K: drive that might explain why folder creation sometimes fails?

**Comment by boocs** [comment]
I tried different hard drives. 

Instead of using my UE4 projects, I went barebones with just a folder and some blank C++ files. It still happens.

I also tried a barebones user config and it still happens.
    "C_Cpp.intelliSenseCacheSize": 0,
    "C_Cpp.default.intelliSenseMode": "msvc-x64",
    "C_Cpp.intelliSenseCachePath": "k:\\\\.vscode\\\\cache\\\\default\\\\user",
    "C_Cpp.default.browse.databaseFilename": "i:\\\\.vscode\\\\cache\\\\browse\\\\${workspaceFolderBasename}\\\\user.vs.db"


I tried writing the path a different way with the same result:
    "C_Cpp.default.browse.databaseFilename": "c:/.vscode/cache/browse/${workspaceFolderBasename}/user.vs.db"

**Comment by boocs** [comment]
More testing:

I thought it was the ${workspaceFolderBasename} variable causing the bug but I was wrong. I tried other variables and they acted the same way.

I:\\.vscode\\cache\\browse\\${config:editor.snippetSuggestions}\\user.vs.db
I:\\.vscode\\cache\\browse\\${env:USERNAME}\\user.vs.db

So I tested this:
I:\\.vscode\\cache\\browse\\test\\user.vs.db

And it doesn't work either! It acted the same way. Weird.

**I also saw a log message:** Timed out waiting for tag parser initialization.

**I finally did get it to work by only using one parent folder.** These both worked. 
_note: uppercase letters are just from other testing_

"I:\\\\.vscode\\\\USER.VS.DB"
"I:\\\\${workspaceFolderBasename}\\\\USER.VS.DB"

So for some reason any number of subdirectories affects the creation. If I just add one subdirectory, to the path, it will bug like before.

**Comment by boocs** [comment]
More info:

It always fails the first time on the last sub directory + filename.

So if I use this: 
`I:\\1\\2\\3\\4\\5\\6\\7\\8\\9\\0\\USER.VS.DB`

Only this is created the first time:
`I:\\1\\2\\3\\4\\5\\6\\7\\8\\9`

If I use this:
`I:\\1\\2\\3\\4\\5\\USER.VS.DB`

Only this is created the first time:
`I:\\1\\2\\3\\4`

**Comment by sean-mcmanus** [comment]
Hmm, I'm still not reproing this bug. I don't see what could be causing the bug via code review either.
![image](https://user-images.githubusercontent.com/19859882/97914034-1687b800-1d04-11eb-935c-1a3c5bdcf8cb.png)

**Comment by boocs** [comment]
I tested the setting with the rest of the settings files: Workspace, Folder, and Language and the bug happens with all of them.

**I tried the zip version of VSCode Insiders and the bug is gone**. 

Not sure what that means though.

**Comment by boocs** [comment]
More:

If I go back to the original VSCode version and use --user-data-dir, to specify a new user data dir, the bug is still there.

All extensions disabled except the one.

Also tried a new --extensions-dir

**Comment by sean-mcmanus** [comment]
Yeah, I repro the bug with 1.0.1 now -- it appears to be fixed with 1.1.0-insiders3. Our plan is to release 1.1.0 on Thursday unless we hit some unexpected delay.

**Comment by Jianchaoliang** [comment]
unsubscribe


| |
菅朝梁
|
|
18205601716@163.com
|
签名由网易邮箱大师定制
On 11/4/2020 06:07，Sean McManus<notifications@github.com> wrote：

Yeah, I repro the bug with 1.0.1 now -- it appears to be fixed with 1.1.0-insiders3. Our plan is to release 1.1.0 on Thursday unless we hit some unexpected delay.

**Comment by boocs** [comment]
Fixed with v1.1.0 on VSCode v1.51.0
